### PR TITLE
Ensure audit log permissions are restricted

### PR DESCRIPTION
**Pull request by JAORMX**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

While the apiserver audit options merely use the lumberjack logger in
order to write the appropriate log files, this library has very loose
permissions by default for these files [1]. However, this library will
respect the permissions that the file has, if it exists already. This is
also the most tested scenario in the library [2].

So, let's follow the pattern marked in the library's tests and
pre-create the audit log file with an appropriate mode.

[1] https://github.com/natefinch/lumberjack/blob/v2.0/lumberjack.go#L280
[2] https://github.com/natefinch/lumberjack/blob/v2.0/linux_test.go

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
ACTION REQUIRED: Audit log files are now created with a mode of 0600. Existing file permissions will not be changed. If you need the audit file to be readable by a non-root user, you can pre-create the file with the desired permissions.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 1]
```diff
@@ -515,7 +523,16 @@ func (o *AuditLogOptions) getWriter() io.Writer {
 			MaxSize:    o.MaxSize,
 		}
 	}
-	return w
+	return w, nil
+}
+
+func (o *AuditLogOptions) ensureLogFile() error {
+	mode := os.FileMode(0600)
+	f, err := os.OpenFile(o.Path, os.O_CREATE|os.O_RDWR, mode)
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go:531]: Does this need `os.O_APPEND` to avoid clearing the file?

**Comment by tallclair** [comment]
This should definitely have a release note.

Do we need to do anything special around the rollout to avoid breaking people? I'm inclined to say an `ACTION REQUIRED` release note is sufficient.

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go (hunk 1 above)
Comment (JAORMX) [reply staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: Right! I had missed this. Thanks

**Comment by JAORMX** [comment]
> This should definitely have a release note.
> 
> Do we need to do anything special around the rollout to avoid breaking people? I'm inclined to say an `ACTION REQUIRED` release note is sufficient.

@tallclair  I changed the release note. What do you think?

**Comment by tallclair** [comment]
@JAORMX I made the release note actionable. Can you please check it for accuracy?

**Comment by JAORMX** [comment]
@tallclair thanks for that! The changes are good!

**Comment by JAORMX** [comment]
@tallclair could you revisit this?

**Comment by deads2k** [comment]
If I pre-create the first file, do the historical files take the permissions of the primary file?

@soltysh @sttts This has the potential to break collection and scrapers.  Probably want to be sure permissions carry as expected.

**Comment by JAORMX** [comment]
> If I pre-create the first file, do the historical files take the permissions of the primary file?

Could you rephrase this? I apologize but I didn't quite get the question.

**Comment by deads2k** [comment]
There is a primary file (the one currently being written) and historical files (the ones that are dated and sometimes zipped).  If I set `0640` on the primary file, will the historical files be `0640` or will they be `0600`? Same question when the primary file gets archived and recreated.

**Comment by JAORMX** [comment]
> There is a primary file (the one currently being written) and historical files (the ones that are dated and sometimes zipped). If I set `0640` on the primary file, will the historical files be `0640` or will they be `0600`? Same question when the primary file gets archived and recreated.

Thanks for the clarification! the `os.Open` call will only set the file mode if it's not already set. So if there was a pre-existing file with `0640` it will retain that mode. The rotation is handled elsewhere (by the lumberjack library), and it preserves the mode that the file started with. So if you had `0640`, the rotation will keep that mode in the rotated files.

**Comment by soltysh** [comment]
> > 
> 
> @deads2k thanks for the clarification! the `os.Open` call will only set the file mode on file creation; not when appending to it. So if there was a pre-existing file with `0640` it will retain that mode. The rotation is handled elsewhere (by the lumberjack library), and it preserves the mode that the file started with. So if you had `0640`, the rotation will keep that mode in the rotated files.

Is there a chance to enforce the new mode even for rotation?

**Comment by JAORMX** [comment]
> > > 
> > 
> > 
> > @deads2k thanks for the clarification! the `os.Open` call will only set the file mode on file creation; not when appending to it. So if there was a pre-existing file with `0640` it will retain that mode. The rotation is handled elsewhere (by the lumberjack library), and it preserves the mode that the file started with. So if you had `0640`, the rotation will keep that mode in the rotated files.
> 
> Is there a chance to enforce the new mode even for rotation?

Not from here, that would need to be an enhancement to the [lumberjack library](https://github.com/natefinch/lumberjack/blob/v2.0/lumberjack.go). I know in OpenShift this is addressed by doing `chmod` [on startup](https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/bindata/v4.1.0/kube-apiserver/pod.yaml#L26) to all audit files, that way one can ensure that they have a secure mode.
